### PR TITLE
Let MonoReachable traversal evaluate BinOps

### PR DESCRIPTION
**Pull request by saethlin**
This fixes the first bug in https://github.com/rust-lang/rust/issues/129282

I don't know what the right organization is for the code here, but something probably has to get rearranged because this traversal now depends on `rustc_const_eval` so that it can use an `InterpCx` internally, so it can't live in `rustc_middle`.

r? @ghost

### CODE REVIEW

**Comment by saethlin** [comment]
> Pondering: Can we just tell people that they have to write this in a const block if they want this?
> (AKA "we should have const if

I'm not interested in changing or adding to the language. This seems like a useful optimization for the compiler to do, so I'm trying to see how that turns out.

I think the implementation ends up convoluted because of some deeper issues we have that I'm not going to solve. MIR is really not easy to ad-hoc evaluate, and also we don't have any post-mono MIR opts. Post-mono GVN for example would fix this. I'm pasting most of the impl out of GVN to do this.

@bors try @rust-timer queue

File: compiler/rustc_const_eval/src/lib.rs [hunk 1]
```diff
@@ -51,7 +51,107 @@ pub fn provide(providers: &mut Providers) {
...
+
 /// `rustc_driver::main` installs a handler that will set this to `true` if
 /// the compiler has been sent a request to shut down, such as by a Ctrl-C.
 /// This static lives here because it is only read by the interpreter.
 pub static CTRL_C_RECEIVED: AtomicBool = AtomicBool::new(false);
+
+/// If this basic block ends with a [`TerminatorKind::SwitchInt`] for which we can evaluate the
+/// dimscriminant in monomorphization, we return the discriminant bits and the
+/// [`SwitchTargets`], just so the caller doesn't also have to match on the terminator.
+pub fn try_const_mono_switchint<'a, 'tcx>(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/lib.rs:68]: Please find a better place for this function -- the crate root is not a good location for any code.

Worst-case, put it in `util`.

**Comment by RalfJung** [comment]
> I'm not interested in changing or adding to the language.

It shouldn't be needed to change the language. This is stable:
```rust
  if const { SIZE < 4096 } {
    let arr = [0u8; SIZE];
    std::hint::black_box(&arr);
  } else {
    let vec = vec![0u8; SIZE];
    std::hint::black_box(&vec);
  }
```

If that works, IMO there's no sufficient justification for piling on hacks like this.

**Comment by RalfJung** [comment]
Seems like that [doesn't work](https://github.com/rust-lang/rust/issues/129282#issuecomment-2298627070). We should figure out why. I don't think we want to start integrating general optimizations into the monomorphization step -- if we decide we need post-mono optimizations on the MIR level, we should do it properly and generate MIR for that, not embed a bunch of hard-to-audit on-the-fly transformations hidden in various other passes.

**Comment by RalfJung** [comment]
> We already know why: https://github.com/rust-lang/rust/pull/129283

So #129283 plus using `const { ... }` is enough to fix https://github.com/rust-lang/rust/issues/129282? IMO that's enough. It's a debug build, after all, it's not even asking to be optimized.

**Comment by saethlin** [comment]
> there's no sufficient justification for piling on hacks like this.

This PR is an experiment. Unfortunately I can't collect perf results without posting a PR. I've set the status to "experimental" instead of "waiting on author" is that more clear?

**Comment by RalfJung** [comment]
Fair. :)

File: compiler/rustc_const_eval/src/lib.rs [hunk 2]
```diff
@@ -51,7 +51,107 @@ pub fn provide(providers: &mut Providers) {
...
+        }
+        Operand::Move(place) | Operand::Copy(place) => place,
+    };
+
+    // MIR for `if false` actually looks like this:
+    // _1 = const _
+    // SwitchInt(_1)
+    //
+    // And MIR for if intrinsics::ub_checks() looks like this:
+    // _1 = UbChecks()
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_const_eval/src/lib.rs:102]: No, that would be really odd.

We could consider making `UbChecks` a const (giving it a DefId), similar to how we are making nullary intrinsics effectively consts. However, `UbChecks` is not actually a const: its value depends on `tcx.sess` which can have different values in different crates across the crate graph!

**Comment by saethlin** [comment]
The merge conflicts here are a mess, so I don't plan on fixing them. I think the way to get such an operation merged is by doing post-mono GVN or something like that.

**Pull request closed without merging**